lib/pdfkit/source.rb

```diff
@@ -1,23 +1,45 @@
 class PDFKit
   class Source
+    SOURCE_FROM_STDIN = '-'
+
     def initialize(url_file_or_html)
       @source = url_file_or_html
     end
-    
+
     def url?
-      @source.is_a?(String) && @source.match(/\Ahttp/)
+      @is_url ||= @source.is_a?(String) && @source.match(/\Ahttp/)
     end
-    
+
     def file?
-      @source.kind_of?(File)
+      @is_file ||= @source.kind_of?(File)
     end
-    
+
     def html?
-      !(url? || file?)
+      @is_html ||= !(url? || file?)
     end
-    
+
+    def to_input_for_command
+      if file?
+        @source.path
+      elsif url?
+        %{"#{shell_safe_url}"}
+      else
+        SOURCE_FROM_STDIN
+      end
+    end
+
     def to_s
       file? ? @source.path : @source
     end
+
+    private
+
+    def shell_safe_url
+      url_needs_escaping? ? URI::escape(@source) : @source
+    end
+
+    def url_needs_escaping?
+      URI::decode(@source) == @source
```

Comment:
So if we're quoting the URI, this still won't work.

```rb
1.9.3-p448 :001 > url = 'https://google.com/search?q="foobarbogus"'
 => "https://google.com/search?q=\"foobarbogus\""
1.9.3-p448 :002 > require 'uri'
 => true
1.9.3-p448 :003 > URI::decode(url)
 => "https://google.com/search?q=\"foobarbogus\""
```

That's a simple example that would allow someone to use something like

```rb
1.9.3-p448 :001 > url = 'https://google.com/search?q="useradd -p secrete hacked"'
 => "https://google.com/search?q=\"useradd -p secrete hacked\""
1.9.3-p448 :002 > require 'uri'
 => true
1.9.3-p448 :003 > URI::decode(url)
 => "https://google.com/search?q=\"useradd -p secrete hacked\""
```

We **really** need to first parse the URI to make sure it's valid before checking if the decoded value is the same as the source. If it's not valid then the decoded URI being equal to the original is just a worthless check.